File: tree-js/simple-node.js
const Resource = require('./resource');


class SimpleNode {
    constructor () {
        this.res = new Resource(this.onResourceReady.bind(this));
        // ready flag
        this.ready = false;
        // node ID
        this.id = 'xx-xxx-xx'.replace(/x/g, () => Math.round(Math.random() * 15).toString(16));
        // stack for children
        this.childs = [];
    }

    /**
     * an event handler called when the resource becomes ready, should not be called directly
     * @param {string} value the resource payload (can be left unused for the scope of the assignment)
     */
    onResourceReady (value) {
        // when finished we set ready flag to true
        this.ready = true;
    }

    /**
     * adds child node as a child of this node
     * @param {SimpleNode} child the child to add
     */
    addChild (child) {
        // insert new nodes
        this.childs.push(child);
    }

    /**
     * removes an existing child from this node
     * @param {SimpleNode} child the child to remove
     */
    removeChild (child) {
        // remove child by ID
        this.childs = this.childs.filter(
            ( _child ) => {
                return _child.id !== child.id
            }
        );
    }

    /**
     * checks whether the resources of this node and its subtree are all ready
     * @returns {boolean} are all resources ready
     */
    isReady () {
        // check if node has children
        if (this.childs.length > 0) {
            // run in loop and if all are finished return true
            return this.childs.every(
                (child)=>{
                    return child.isReady(); 
                }
            );
        } else {
            // return ready flag
            return this.ready;
        }
    }
}

module.exports = SimpleNode;

File: tree-py/simple_node.py
from resource import Resource


class SimpleNode:
    def __init__(self):
        self.res = Resource(self.on_resource_ready)

        # implement

    def on_resource_ready(self, value):
        '''
        an event handler called when the resource becomes ready, should not be called directly

        Args:
            value (str): the resource payload (can be left unused for the scope of the assignment)
        '''
        # implement
        pass

    def add_child(self, child):
        '''
        adds child node as a child of this node

        Args:
            child (SimpleNode): the child to add
        '''
        # implement
        pass

    def remove_child(self, child):
        '''
        removes an existing child from this node

        Args:
            child (SimpleNode): the child to add
        '''
        # implement
        pass

    def is_ready(self):
        '''
        checks whether the resources of this node and its subtree are all ready

        Returns:
            bool: are all resources ready
        '''
        # implement
        pass
